### DISCUSSION TITLE
Benifit from AI hardware

### DISCUSSION THREAD

**Discussion by mircodaniel** [official-stockfish/Stockfish#5363] [Q&A] [answered]
Will user with AI hardware (e.g. NPU) somehow benifit using Stockfish NNUE?

**Comment by bftjoe** [comment]
I'm not an expert on this, but this kind of hardware probably would bring better gains to leela or a much larger net than what stockfish currently uses.

**Comment by Sopel97** [comment, marked as answer]
Doesn't look promising with the way manufacturers are hiding the internals and only exposing very high level API. We need instruction-level access to do anything worthwhile. There is also a question whether these NPUs can accelerate vector x matrix operations, or whether they just target matrix x matrix, like for example AMX, which we measured to be worse than AVX2.

**Comment by mstembera** [comment]
@Sopel97 What code was used for the AMX measurement?  I always thought AffineTransformSparseInput::propagate() could benefit from AMX.

**Comment by Sopel97** [comment]
@mstembera https://github.com/Sopel97/Stockfish/commits/amx_v1/, https://github.com/Sopel97/Stockfish/commits/amx_v2/

benchmarks from vondele, I assume it's okay to post this after more than a year

```
Performance
===========

master amx_v* branches

singlethread `bench 128 1 15 default depth NNUE`

bench-876906965b8d-x86-64-vnni512-singlethread-8:Nodes/second    : 1702401
bench-876906965b8d-x86-64-vnni256-singlethread-57:Nodes/second    : 1457396
bench-876906965b8d-x86-64-avx512-singlethread-15:Nodes/second    : 1628402
bench-876906965b8d-x86-64-bmi2-singlethread-82:Nodes/second    : 1692177
bench-876906965b8d-x86-64-avx2-singlethread-13:Nodes/second    : 1668221
bench-2939752feeeb-x86-64-amx-singlethread-9:Nodes/second    : 1324282
bench-0ae8b5781789-x86-64-amx-singlethread-2:Nodes/second    : 1324282

224 threads. startpos depth 40, 16GB

stockfish-0ae8b5781789-x86-64-amx nps 119238124 
stockfish-876906965b8d-x86-64-vnni256 nps 137106606 
stockfish-876906965b8d-x86-64-bmi2 nps 138381301 
stockfish-876906965b8d-x86-64-vnni512 nps 139242129 
```
